### PR TITLE
[libraries] Move library tests Feature Switches defaults to Functional tests

### LINKED ISSUE #52645: WebSockets test failures on Android
As per https://github.com/dotnet/runtime/pull/52052#issuecomment-838655841, creating an issue to track WebSockets test failures on Android and put them under ActiveIssue.

`ubuntu.1804.amd64.android` - x86, x64

- System.Net.WebSockets.Tests.WebSocketDeflateTests.WebSocketWithoutDeflateShouldThrowOnCompressedMessage
```
Assert.Equal() Failure
          ↓ (pos 0)
Expected: The WebSocket received compressed frame w···
Actual:   net_Websockets_PerMessageCompressedFlagWh···
          ↑ (pos 0)
```

- System.Net.WebSockets.Tests.WebSocketDeflateTests.ReceiveUncompressedMessageWhenCompressionEnabled
```
Assert.Equal() Failure
          ↓ (pos 0)
Expected: The WebSocket received compressed frame w···
Actual:   net_Websockets_PerMessageCompressedFlagWh···
          ↑ (pos 0)
```

- System.Net.WebSockets.Tests.WebSocketDeflateTests.ReceiveInvalidCompressedData
```
Assert.Equal() Failure
          ↓ (pos 0)
Expected: The message was compressed using an unsup···
Actual:   ZLibUnsupportedCompression
          ↑ (pos 0)
```

### PR DESCRIPTION
**Pull request by mdh1418**
Fixes #52645 

In the work to reduce library testing infrastructure size, `EnableAggressiveTrimming` had been added in addition to a few properties from https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming-options#trimming-framework-library-features. 

These properties were not conditioned under `EnableAggressiveTrimming` or made library testing specific, so they probably attributed to https://github.com/dotnet/runtime/issues/48027, which had been resolved by https://github.com/dotnet/runtime/pull/48429 using `DefaultBlazorWASMFeatureSwitches`.

In the work to move WASM AOT test building to helix, some of the properties were modified https://github.com/dotnet/runtime/pull/48226, probably to get tests buliding and passing.

In a similar vein, Xamarin Android had set switch defaults for runtime libraries also to reduce size https://github.com/xamarin/xamarin-android/pull/5337, and this work had been extended to `dotnet/runtime` via 
 https://github.com/dotnet/runtime/pull/49005 and https://github.com/dotnet/runtime/pull/49635 because it was thought that because Browser wasm had them in `tests.wasm.targets` that they would be needed for Android and iOS.

Taking the above events into account, the properties were originally intended to reduce library testing size when trimming (so when `EnableAggressiveTrimming=true`) but weren't conditioned as such, and led to the other events. 

This PR looks to:

- Remove ActiveIssue 52645 attributes

- Set the trimming framework features properties directly in the Functional tests project properties and model them after [Xamarin Android](https://github.com/radekdoulik/xamarin-android/blob/038c42385af103957e3e9bbee21a53f1dd400e1a/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.ILLink.targets#L13-L23), [Xamarin Mac/iOS](https://github.com/xamarin/xamarin-macios/blob/aad913db8673637150af01b13e76605966b8f14a/dotnet/targets/Xamarin.Shared.Sdk.targets#L118-L134), and [WASM](https://github.com/dotnet/sdk/blob/c90ef5e103e5c961e40acd6ab1b1622353cc2965/src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets#L57-L63)

- Set the trimming framework features properties in tests.mobile.targets to the values previously found in tests.wasm.targets

- Remove `UseDefault*FeatureSwitches`

### CODE REVIEW

**Comment by filipnavara** [comment]
Why only for Android and not for iOS? Is there some background on why it was used for tests in the first place?

**Comment by steveisok** [comment]
> Why only for Android and not for iOS? Is there some background on why it was used for tests in the first place?

We wanted standard defaults for the functional tests and I think it got mixed with the libraries tests.  Fair point re: iOS.  I agree.

**Comment by radical** [comment]
> We wanted standard defaults for the functional tests and I think it got mixed with the libraries tests. 

This should be added to the PR description too.

**Comment by filipnavara** [comment]
LGTM

**Comment by radical** [comment]
> We wanted standard defaults for the functional tests and I think it got mixed with the libraries tests.

1. Is `$(UseDefaultAndroidFeatureSwitches)` ever !="true"? Just wondering if we need these properties at all.
2. If these were added just for functional tests, and got added for libraries by mistake, then can/should the equivalent thing in `tests.wasm.targets` be removed too?

cc @akoeplinger

**Comment by filipnavara** [comment]
> Is $(UseDefaultAndroidFeatureSwitches) ever !="true"? Just wondering if we need these properties at all.

No. Same for iOS.

**Comment by mdh1418** [comment]
The default switches properties are set to false in the scope of `/eng/testing/linker/SupportFiles/Directory.Build.props`
https://github.com/dotnet/runtime/blob/84b7c672427b98e70c35c58232ff573625c96f6a/eng/testing/linker/SupportFiles/Directory.Build.props#L4-L6
@MaximLipnin I'm not exactly sure why they're needed, do you know in more detail?

I'll move the Browser wasm equivalent from `tests.mobile.targets` to `src/tests/FunctionalTests/Directory.Build.props`

**Comment by radical** [comment]
You can also drop the 3 properties from `tests.mobile.targets` (https://github.com/mdh1418/runtime/blob/remove_use_system_resource_keys_from_libraries/eng/testing/tests.mobile.targets#L9-L11).

**Comment by radical** [comment]
.. and move them to the function tests csproj!

**Comment by mdh1418** [comment]
Ah right! Totally missed that, thanks :)

**Comment by MaximLipnin** [comment]
> The default switches properties are set to false in the scope of `/eng/testing/linker/SupportFiles/Directory.Build.props`
> https://github.com/dotnet/runtime/blob/84b7c672427b98e70c35c58232ff573625c96f6a/eng/testing/linker/SupportFiles/Directory.Build.props#L4-L6
> 
> 
> @MaximLipnin I'm not exactly sure why they're needed, do you know in more detail?

I added ios/android props analogously to the browser one. If it's not necessary now we could remove them.

**Comment by mdh1418** [comment]
I think I discovered the series of events that led to the properties as of current, and updated the PR description. Can I get another 👀 over if it makes sense? @filipnavara @radical @MaximLipnin

**Comment by radical** [comment]
> I think I discovered the series of events that led to the properties as of current, and updated the PR description. Can I get another 👀 over if it makes sense? @filipnavara @radical @MaximLipnin

Awesome! Thank you for getting that whole history figured out 👍

**Review by radical — COMMENTED** [review]
This does explicitly set some new properties, for the wasm case, which might be fine. Let's wait to see how the builds go.

**Comment by filipnavara** [comment]
Thanks for putting together the links.

For reference, here are the defaults on the Xamarin.Mac/iOS side since the Xamarin.Android ones are already linked in the description:
https://github.com/xamarin/xamarin-macios/blob/aad913db8673637150af01b13e76605966b8f14a/dotnet/targets/Xamarin.Shared.Sdk.targets#L118-L134
Same for the WASM:
https://github.com/dotnet/sdk/blob/c90ef5e103e5c961e40acd6ab1b1622353cc2965/src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets#L57-L63

I'm still somewhat torn on what is the intended setup. On one hand keeping some of the switches "on" may reduce the size of the packages that are produced and transferred to Helix and in turn save some time and bandwidth. It also more closely mirrors the defaults that customer will get when targeting `net6.0-ios` / `net6.0-android`. On the other hand it causes compatibility errors with some of the tests and the environment they expect and it also overrides runtimeconfig.template.json (https://github.com/dotnet/runtime/pull/53172#issue-651301757) which caught me off guard (and perhaps should be tracked as dotnet/sdk issue?).

I'd certainly value some other opinion on what do we really want to test here since there's no clear answer to that. Is the intent for functional tests to mirror the options customers get? Are the unit tests intentionally run in an environment that is unified across platforms but not matching the defaults customers get? Are there actually some tests specifically for these options and whether they are honoured properly?

**Comment by steveisok** [comment]
The intent of the functional tests are to closely mirror what our customers will get w/ xamarin android and ios.  I had forgotten we originally brought these defaults up for our wasm AOT testing efforts (where a smaller app really matters).  

As for the libraries tests, the problem with applying centralized defaults in tests.mobile.targets is that there are no opportunities for the test project to override them.  Also, I would rather not try to deviate from having the test project be in control of its own configuration.

**Comment by filipnavara** [comment]
> the problem with applying centralized defaults in tests.mobile.targets is that there are no opportunities for the test project to override them

I started to introduce a mechanism to override them on per-test-project basis by adding the same `Conditon="'$(XXX)' == ''"` attribute to each of the options.

I am fine with keeping the defaults different for functional tests (that finally run as part of CI) and the unit tests since it makes some things easier. Just wanted to point out that there are some downsides to it too.

That leaves the last question, is there a test coverage somewhere (dotnet/sdk? linker?) specifically for testing the options themselves? For example, I would expect tests that explicily expect 1) disabled UTF-7 encoding 2) exception messages having the resource key instead of localized string 3) disabled binary serialization.

**Comment by mdh1418** [comment]
Thanks for the input! 

> The intent of the functional tests are to closely mirror what our customers will get w/ xamarin android and ios
Because of this, I decided to add the Xamarin Android, Xamarin Mac/iOS, and Wasm feature defaults into the Functional tests project properties.

> I started to introduce a mechanism to override them on per-test-project basis by adding the same Conditon="'$(XXX)' == ''" attribute to each of the options as there is in the workloads/SDKs.
I copied this scheme for the properties set in `tests.mobile.targets`

File: src/tests/FunctionalTests/Directory.Build.props [hunk 1]
```diff
@@ -7,5 +7,44 @@
...
+		<EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization Condition="'$(EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization)' == ''">false</EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization>
+		<EnableUnsafeBinaryFormatterSerialization Condition="'$(EnableUnsafeBinaryFormatterSerialization)' == ''">false</EnableUnsafeBinaryFormatterSerialization>
+		<EnableUnsafeUTF7Encoding Condition="'$(EnableUnsafeUTF7Encoding)' == ''">false</EnableUnsafeUTF7Encoding>
+		<EventSourceSupport Condition="'$(EventSourceSupport)' == ''">false</EventSourceSupport>
+		<HttpActivityPropagationSupport Condition="'$(HttpActivityPropagationSupport)' == ''">false</HttpActivityPropagationSupport>
+		<InvariantGlobalization Condition="'$(TargetOS)' == 'MacCatalyst' And '$(InvariantGlobalization)' == ''">true</InvariantGlobalization>
+		<StartupHookSupport Condition="'$(StartupHookSupport)' == ''">false</StartupHookSupport>
+		<UseSystemResourceKeys Condition="'$(UseSystemResourceKeys)' == ''">true</UseSystemResourceKeys>
+		<UseNativeHttpHandler Condition="'$(UseNativeHttpHandler)' == ''">true</UseNativeHttpHandler>
```
Comment (radical) [inline src/tests/FunctionalTests/Directory.Build.props:46]: tabs -> spaces

Comment (mdh1418) [reply src/tests/FunctionalTests/Directory.Build.props:46]: Thanks for the catch :)

File: src/tests/FunctionalTests/Directory.Build.props [hunk 2]
```diff
@@ -7,5 +7,44 @@
...
+    <InvariantGlobalization Condition="'$(InvariantGlobalization)' == ''">false</InvariantGlobalization>
+    <UseSystemResourceKeys Condition="'$(UseSystemResourceKeys)' == ''">true</UseSystemResourceKeys>
+  </PropertyGroup>
+
+  <!-- Modeled after Xamarin.Shared.Sdk.targets -->
+  <PropertyGroup Condition="'$(TargetOS)' == 'iOS' or '$(TargetOS)' == 'tvOS' or '$(TargetOS)' == 'MacCatalyst'">
+		<_BundlerDebug Condition="'$(_BundlerDebug)' == '' And '$(Configuration)' == 'Debug'">true</_BundlerDebug>
+		<_BundlerDebug Condition="'$(_BundlerDebug)' == ''">false</_BundlerDebug>
+
+    <DebuggerSupport Condition="'$(DebuggerSupport)' == ''">$(_BundlerDebug)</DebuggerSupport>
```
Comment (radical) [inline src/tests/FunctionalTests/Directory.Build.props]: Is `$(_BundlerDebug)` going to be set by anyone else? If not, then you can fold this into `DebuggerSupport` itself, like:

```xml
<DebuggerSupport Condition="'$(DebuggerSupport)' == '' and '$(Configuration)' == 'Debug'">true</DebuggerSupport>
<DebuggerSupport Condition="'$(DebuggerSupport)' == ''">false</DebuggerSupport>
```

File: src/tests/FunctionalTests/Directory.Build.props [hunk 3]
```diff
@@ -7,5 +7,44 @@
     <IsFunctionalTest>true</IsFunctionalTest>
   </PropertyGroup>
 
+  <!-- Modeled after Microsoft.Android.Sdk.ILLink.targets -->
```
Comment (radical) [inline src/tests/FunctionalTests/Directory.Build.props:10]: might be useful to link the file. same for the other targets files.

File: src/tests/FunctionalTests/Directory.Build.props (hunk 2 above)
Comment (mdh1418) [reply src/tests/FunctionalTests/Directory.Build.props]: I suppose not, Xamarin.Shared.Sdk.targets wouldn't affect this project somehow right? Thanks for the catch.

File: src/tests/FunctionalTests/Directory.Build.props [hunk 4]
```diff
@@ -7,5 +7,44 @@
...
+    <EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization Condition="'$(EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization)' == ''">false</EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization>
+    <EnableUnsafeBinaryFormatterSerialization Condition="'$(EnableUnsafeBinaryFormatterSerialization)' == ''">false</EnableUnsafeBinaryFormatterSerialization>
+    <EnableUnsafeUTF7Encoding Condition="'$(EnableUnsafeUTF7Encoding)' == ''">false</EnableUnsafeUTF7Encoding>
+    <EventSourceSupport Condition="'$(EventSourceSupport)' == ''">false</EventSourceSupport>
+    <HttpActivityPropagationSupport Condition="'$(HttpActivityPropagationSupport)' == ''">false</HttpActivityPropagationSupport>
+    <InvariantGlobalization Condition="'$(TargetOS)' == 'MacCatalyst' And '$(InvariantGlobalization)' == ''">true</InvariantGlobalization>
+    <StartupHookSupport Condition="'$(StartupHookSupport)' == ''">false</StartupHookSupport>
+    <UseSystemResourceKeys Condition="'$(UseSystemResourceKeys)' == ''">true</UseSystemResourceKeys>
+    <UseNativeHttpHandler Condition="'$(UseNativeHttpHandler)' == ''">true</UseNativeHttpHandler>
```
Comment (radical) [inline src/tests/FunctionalTests/Directory.Build.props:46]: super nit: reorder to have the common properties first, then the extra apple ones after a new line. feel free to ignore!

Comment (radical) [reply src/tests/FunctionalTests/Directory.Build.props:46]: I meant, the properties that are in the other two blocks (android, and browser), just to make it easier to read, and compare.

File: src/tests/FunctionalTests/Directory.Build.props (hunk 3 above)
Comment (mdh1418) [reply src/tests/FunctionalTests/Directory.Build.props]: Done!

**Comment by mdh1418** [comment]
The failing lanes seem to be unrelated. 
Android x64 RuntimeTests is a PackageInstallationError
The other lane is related to the installer for CoreCLR on Linux, where this PR is relevant for library tests.

**Review by steveisok — APPROVED** [review]

**Review by radical — APPROVED** [review]

**Pull request merged**